third-culture-software/bhima
# Increase the precision of the exchange rate

This PR changes the data type for the exchange rate to DOUBLE precision.  This ensures that exchange rate conversions do not loose precision.

It also adds a hint to the exchange rate entry modal:
![image](https://user-images.githubusercontent.com/62145/146419077-98895096-dced-4259-98cf-313267081423.png)

Closes https://github.com/IMA-WorldHealth/bhima/issues/6195

**TESTING**
- Use a database that uses FC as the enterprise currency (eg, Vanga - NOT bhima_test)
- Make sure to apply the migrations.sql in this PR!
- Set the exchange rate for USD :  2030 FC/$ (enter all digits of 1/2030 = 0.00049261084)
- Create a purchase order
   - Select USD as the currency 
   - Select any stock item
   - Enter the unit price as:   3.0
   - Set the quantity to 100000
   - Save the purchase order.  Note the total order is:  $300,000.00
- Confirm the purchase order (via the Purchase Order Registry action menu for that PO)
- Enter stock for for that PO.
  - When entering/creating the LOT, note that the full unit price is shown:  3
  - Accept all 100000 items
  - Verify the Entry document shows $300,000.00 as the total
- Go to the Stock Entries Report page
  - Check the 'Include entries from purchases' and 'Show details' items 
  - View the report in USD and verify the price for that order is $300,000.00
  - View the report in FC.  It should show:  300000 * 2030 = 6.090.000.000,00 FC

NOTE: Before this fix, the last value would have been incorrect because the fractional value of 1/2030 is limited to 8 digits (including leading zeros):

      300000 / 0.00049261 = 609.001.035.302 FC

NOTE: If you want to see the before/after effect of this change. 

- Use the Vanga dataset.  
- Reload the database and the migrations.sql.  Do NOT change the exchange rate yet.  The FC/USD rate should be 2030.   
- Follow the steps above (but not setting a new exchange rate first).  
- Then set the USD/FC exchange rate to its full value: 

     1 / 2030.0 = 0.00049261083744
- Repeat the procedure above.  Then in the Shoe Entries Report you should see both the values with and without this correction.
